File: libjava/java/security/acl/Group.java
/* Group.java -- Represents a group of Principals
   Copyright (C) 1998 Free Software Foundation, Inc.

This file is part of GNU Classpath.

GNU Classpath is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

GNU Classpath is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with GNU Classpath; see the file COPYING.  If not, write to the
Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
02111-1307 USA.

As a special exception, if you link this library with other files to
produce an executable, this library does not by itself cause the
resulting executable to be covered by the GNU General Public License.
This exception does not however invalidate any other reasons why the
executable file might be covered by the GNU General Public License. */

package java.security.acl;

import java.security.Principal;
import java.util.Enumeration;

/**
 * This interface represents a group of <code>Principals</code>.  Note that
 * since this interface extends <code>Principal</code>, a <code>Group</code>
 * can be used where ever a <code>Principal</code> is requested.  This
 * includes arguments to the methods in this interface.
 *
 * @version 0.0
 *
 * @author Aaron M. Renn (arenn@urbanophile.com)
 */
public interface Group
{
  /**
   * This method adds a new <code>Principal</code> to this group.
   *
   * @param user The new <code>Principal</code> to add
   *
   * @return <code>true</code> if the user was successfully added or <code>false</code> if the user is already a member
   */
  public abstract boolean addMember(Principal user);

  /**
   * This method deletes a member from the group.
   *
   * @param user The <code>Principal</code> to delete
   *
   * @return <code>true</code> if the user was successfully deleted or <code>false</code> if the user is not a member of the group
   */
  public abstract boolean removeMember(Principal user);

  /**
   * This method tests whether or not a given <code>Principal</code> is a
   * member of this group.
   *
   * @param user The <code>Principal</code> to test for membership
   *
   * @return <code>true</code> if the user is member, <code>false</code> otherwise
   */
  public abstract boolean isMember();

  /**
   * This method returns a list of all members of the group as an 
   * <code>Enumeration</code>.
   *
   * @return The list of all members of the group
   */
  public abstract Enumeration members();
}

File: libgrust/libproc_macro/group.rs
// Copyright (C) 2023-2024 Free Software Foundation, Inc.
//
// This file is part of the GNU Proc Macro Library.  This library is free
// software; you can redistribute it and/or modify it under the
// terms of the GNU General Public License as published by the
// Free Software Foundation; either version 3, or (at your option)
// any later version.

// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// Under Section 7 of GPL version 3, you are granted additional
// permissions described in the GCC Runtime Library Exception, version
// 3.1, as published by the Free Software Foundation.

// You should have received a copy of the GNU General Public License and
// a copy of the GCC Runtime Library Exception along with this program;
// see the files COPYING3 and COPYING.RUNTIME respectively.  If not, see
// <http://www.gnu.org/licenses/>.

use bridge;
use std::fmt;
use Span;
use TokenStream;

/// Describes how a sequence of token trees is delimited.
#[repr(C)]
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum Delimiter {
    /// The sequence is delimited by a parenthesis `(...)`.
    Parenthesis,
    /// The sequence is delimited by a brace `{...}`.
    Brace,
    /// The sequence is delimited by a bracket `[...]`.
    Bracket,
    /// Invisible delimiter to preserve operator priority.
    None,
}

/// A delimited token stream.
#[derive(Clone)]
pub struct Group(pub(crate) bridge::group::Group);

impl Group {
    /// Creates a new `Group`.
    ///
    /// # Arguments
    ///
    /// * `delimiter` - The delimiter surrounding the inner [`TokenStream`].
    /// * `stream` - The tokenstream for this `Group`.
    pub fn new(delimiter: Delimiter, stream: TokenStream) -> Self {
        Group(bridge::group::Group::new(delimiter, stream.0))
    }

    /// Get the delimiter of the `Group`.
    pub fn delimiter(&self) -> Delimiter {
        self.0.delimiter()
    }

    /// Get the stream of the `Group`.
    ///
    /// # Note
    ///
    /// The returned stream does not include the delimiters of this group.
    pub fn stream(&self) -> TokenStream {
        TokenStream(self.0.stream())
    }

    /// Get the span for the delimiters of this token stream, spanning the
    /// entire group.
    pub fn span(&self) -> Span {
        Span(self.0.span())
    }

    /// Get the span pointing to the opening delimiter of this `Group`.
    pub fn span_open(&self) -> Span {
        Span(self.0.span())
    }

    /// Get the span pointing to the closing delimiter of this `Group`.
    pub fn span_close(&self) -> Span {
        Span(self.0.span())
    }

    /// Change the span for this `Group`'s delimiters, but not its internal
    /// tokens.
    ///
    /// # Note
    ///
    /// This method will **not** set the span of all the internal tokens spanned
    /// by this group, but rather it will only set the span of the delimiter
    /// tokens at the level of the `Group`.
    pub fn set_span(&mut self, span: Span) {
        self.0.set_span(span.0)
    }
}

impl fmt::Display for Group {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}

impl fmt::Debug for Group {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}
